Some changes/enhancements to Sunburst chart

Crenshinibon opened the pull request:
Hi,

I was actually trying to fix the updating of the sunburst chart (https://github.com/novus/nvd3/issues/1317), but utterly failed.

But along the way I cleaned up some stuff and added some other functionality.

- customizable labels
- tooltip value formatter as parameter
- sort parameter to specify the order of the arcs 

Labels were requested in this issue: https://github.com/novus/nvd3/issues/1284

Crenshinibon commented:
![selection_005](https://cloud.githubusercontent.com/assets/69277/10998277/4e62f482-8494-11e5-8e53-65ec0ddf11e0.png)

Crenshinibon commented on src/models/sunburst.js:
```diff
@@ -60,16 +102,6 @@ nv.models.sunburst = function() {
-            };
-            chart.container = this;
-
-
```
I removed this, because it didn't solve the problem. In fact it wasn't called ...

Crenshinibon commented on src/models/sunburst.js:
```diff
@@ -144,48 +239,51 @@ nv.models.sunburst = function() {
+            //     } else {
+            //       return tween;
+            //     }
+            //   }
```
I commented out the function because it wasn't called, but probably should in case the updating is triggered.

liquidpele replied:
Please don't comment it out

Crenshinibon replied:
Should I better delete it than? 

I guess, someone with a better grasp of the whole update-matter of d3 should have a look at it ... the way it was implemented could have never worked. 

Or maybe I should have another look at the implementation of the original example where this chart was inspired from.

liquidpele replied:
Maybe...  I originally tried to copy the example here when I made it, so the updating thing I think was more about changing the way the slices were sized.... 

http://bl.ocks.org/kerryrodden/477c1bfb081b783f80ad

liquidpele commented on examples/documentation.html:
```diff
@@ -899,6 +899,26 @@ <h3 class="option"><span class="expand all">[ <a href="javascript:$('#optionlist
+            default: "function(d1, d2){return d1.name > d2.name;}",
+            examples: [function(d1, d2){return d1.value > d2.value;}]
+        },
+        tooltipValueFormatter: {
```
You could already do this by simply doing chart.tooltip.valueFormatter(..), adding a duplicate way to do it here will cause confusion.

Crenshinibon replied:
I wasn't aware of that options. I will remove it.

liquidpele commented on examples/documentation.html:
```diff
@@ -899,6 +899,26 @@ <h3 class="option"><span class="expand all">[ <a href="javascript:$('#optionlist
+            default: "false",
+            examples: [true, false]
+        },
+        labelText: {
```
DiscreteBarChart has a similar thing, called valueFormat:

https://nvd3-community.github.io/nvd3/examples/documentation.html#discreteBarChart

We should use the same option name if it does the same thing.

liquidpele replied:
hmm... pieChart has labelFormat 

https://nvd3-community.github.io/nvd3/examples/documentation.html#pieChart

that's probably better actually, since this chart is more related to pieChart.

Crenshinibon replied:
I will change it to the pieChart wording than. 'valueFormat' is working differently. You pass it a formatting string, e.g. ',.2f' (which isn't very descriptive, imho). Here, you pass it a function, which is called with the data and it expects that the function returns the formatted string. I will look at the PieChart implementation on monday, as well.

liquidpele replied:
valueFormat takes in a d3 formatter.... you would pass the result of d3.format(',.2f') to it.

Crenshinibon replied:
Just checked the pie ... labelFormat is marked as deprecated:

    // deprecated after 1.7.1
        labelFormat: {get: function(){ return valueFormat;}, set: function(_) {
            valueFormat=_;
            nv.deprecated('labelFormat','use valueFormat instead');
        }},)

And just maps to valueFormat. For my purpose I need some more fine-tuning for the label than just specifying the way the value should be displayed. Should I remove the whole option than or do you suggest another name to make the difference obvious.

liquidpele replied:
What kind of fine tuning??   The labels are just a displayed text value...    if you're referring to the tooltips, you would use the contentGenerator to control the full tooltip content.

liquidpele commented on examples/documentation.html:
```diff
@@ -899,6 +899,26 @@ <h3 class="option"><span class="expand all">[ <a href="javascript:$('#optionlist
             default: "true",
             examples: [true, false]
+        },
+        showLabels: {
```
discreteBarChart has showValues....   we should make options that do the same thing named the same.

liquidpele commented:
This is a lot of changes, I'll have to try it out when I get a chance...

Crenshinibon commented:
I will incorporate your suggestions on monday.

Crenshinibon commented:
Just to let you know, I removed some of the boilerplate code that was wrapping the chart in many 'g' elements. I tried to reduce it to the minimum amount necessary.

Crenshinibon commented:
After a good amount of learning I finally fixed the updating issue: https://github.com/novus/nvd3/issues/1317

I had to introduce a key-function-option that determines how to identify the data elements (which are new, updated, deleted). It defaults to 'name'. 

When the data is updated, the chart zooms out. This was the easiest way to circumvent weird behaviour when there were massive changes to the underlying data.

I rewrote a good amount of the chart and changed along the way the mode-option options. It's now count|value. That makes it more aligning to d3, since value is used there, everywhere.

I finally added a labelThreshold option. It works similar to pie/donut-chart.

liquidpele commented on src/models/sunburstChart.js:
```diff
@@ -90,9 +76,9 @@ nv.models.sunburstChart = function() {
+        evt.series = {
             key: evt.data.name,
-            value: evt.data.size,
+            value: evt.data.value,
```
You changed this, but this will break charts for people already using the chart and expecting size to work.   please do this for backward compatiblity:

    value:  evt.data.value || evt.data.size,

liquidpele replied:
And do the same elsewhere that value is used please :)

liquidpele commented:
Looks pretty good to me besides the one thing, looking forward to merging this to get update working!

Crenshinibon commented:
@liquidpele I hope, I will find tomorrow the time to fix everything up.

Crenshinibon commented:
I'm done so far, but have only checked quickly in my app.

canadian-thackeray commented:
Hi there, I am trying to use the labelFormat and labelThreshold options for the sunburst without any success. I am only getting "label value" ie "vis 432629", and the threshold doesn't seem to reduce the labels when you get to the smaller percentages. Any suggestions on what might be causing this?
